src/common/com/intellij/plugins/haxe/model/HaxeProjectModel.java

```diff
@@ -32,14 +36,30 @@
 import static com.intellij.plugins.haxe.model.HaxeStdTypesFileModel.STD_TYPES_HX;
 
 public class HaxeProjectModel {
+  private static final Key<HaxeProjectModel> HAXE_PROJECT_MODEL_KEY = new Key<>("HAXE_PROJECT_MODEL");
   private final Project project;
-  private final HaxePackageModel stdPackage;
-  private final HaxeSourceRootModel sdkRoot;
+
+  private boolean rootsInitialized = false;
+  private List<HaxeSourceRootModel> roots;
+  private HaxeSourceRootModel sdkRoot;
+  private HaxePackageModel stdPackage;
```

Comment:
I think this could be greatly simplified if the `roots`, `sdkRoot`, and `stdPackage` were in a private class and a variable for that class -- say `rootCache` -- would just be null when they haven't been retrieved.  Then, we could remove the rootsInitialized boolean, and everywhere that you check that flag would just call/return `getRoots().member`.  

`getRoots()` would be the only function to know about `rootCache`, and validate/retrieve them as needed.  The listener would set `rootCache` to `null`; `getRoots()` would validate lazily.